Adding python3 test module for TC_ACL_2_7

j-ororke opened the pull request:
This PR is for the creation of the python3 test module for TC_ACL_2_7 test.
- This uses base branch of Amine's PR #38263, this is a follow-up PR to that PR, once Amine's PR is merged this one can come out of WIP state.
 
#### Testing
WSL - Automated - Test module

Alami-Amine committed:
Modifying how ACL Attribute is written: making it use a non-empty initial ReplaceAll list

Alami-Amine committed:
removing Test modifications on ACL Extensions

Alami-Amine committed:
adding reference to issue in TODO comment

Alami-Amine committed:
integrating comments

Alami-Amine committed:
integrating comments

Alami-Amine committed:
Revert "removing Test modifications on ACL Extensions"

This reverts commit 055af80d6c71e40ddeef081b6046263bc06ef141.

Alami-Amine committed:
Clarifying AttributeDataIB Checkpoint

Alami-Amine committed:
adding Valid CAT values

Alami-Amine committed:
Integrating comments

Alami-Amine committed:
Add comments clarifying AttributeDataIB rollback

Alami-Amine committed:
Integrating comments again

Alami-Amine committed:
change method order

j-ororke committed:
Updating TC_ACL_2_7 test module:
- Adding back in TC_ACL_2_7 python3 test module

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,293 @@
+from mobly import asserts
+
+D_OK_EMPTY = bytes.fromhex('1718')
+D_OK_SINGLE = bytes.fromhex(
```
Can you add a comment that this comes directly from the test plan?

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,293 @@
+            [(0, cfi_attribute)]
+        )
+        f1 = result[0][oc_cluster][cfi_attribute]
+        self.print_step("CurrentFabricIndex F1", str(f1))
```
can you just logging for these? the print_step function adds some stars etc to make it look like test step

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,293 @@
+        # Read CurrentFabricIndex for TH1
+        oc_cluster = Clusters.OperationalCredentials
+        cfi_attribute = oc_cluster.Attributes.CurrentFabricIndex
+        result = await self.th1.ReadAttribute(
```
if you use this: https://github.com/project-chip/connectedhomeip/blob/master/src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py#L1114, you get the success check automatically and don't have to pull the data out of the result dict.

Same below.

j-ororke replied:
This way is much better, sorry, I had just used similar methods to the other ACL tests initially.

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,293 @@
+        # TH1 puts DUT into commissioning mode, TH2 is created and commissions DUT using admin node ID
+        duration = await self.read_single_attribute_check_success(endpoint=0, cluster=Clusters.GeneralCommissioning, attribute=Clusters.GeneralCommissioning.Attributes.BasicCommissioningInfo)
+        params = await self.th1.OpenCommissioningWindow(
+            nodeid=self.dut_node_id, timeout=duration.maxCumulativeFailsafeSeconds, iteration=10000, discriminator=self.discriminator, option=1)
```
maxCumulativeFailsafeSeconds isn't actually the same as the commissioning window limit. Just use 900 - it's not properly exposed anywhere on the device. In theory devices can use other values, in practice they don't. We can fight that fight in the spec if we get a weirdo device.

j-ororke replied:
Ah, ok understood. I had gotten used to gathering the maxCumulativeFailsafeSeconds value and storing it in a var, then using it due to the CADMIN tests, so I have changed this to the 900 seconds now. Will make sure to change this in the other ACL tests that I have created as well.

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,293 @@
+            self.print_step("Write result", str(result))
+            asserts.assert_equal(
+                result[0].Status, Status.Success, "Write should have succeeded")
+        except Exception as e:
```
Catching Exception is dangerous. Exceptions show up as test failures anyway, so you can just remove the try: except: re-assert.

same belwo.

j-ororke replied:
Thank you, I have removed the exceptions from the test and using only the assertions here now as no expected failure scenarios in this test.

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,293 @@
+        except Exception as e:
+            self.print_step("Error reading events for TH2", str(e))
+            asserts.fail(f"Failed to read events for TH2: {e}")
+
```
Not part of the current test plan, but I'd like to suggest adding a cleanup step to remove the extensions and the additional fabric.

j-ororke replied:
Sure, I can happily add that cleanup step into the test here now.

j-ororke committed:
Updating TC_ACL_2_7 python3 test module:
- Added new step to remove TH2 fabric at end of test
- Removed exceptions since not needed due to using asserts, and no expected failure scenarios for this test
- Changed to using static 900 value as the commissioning timeout duration
- Changed to using read_single_attribute_check_success() in place of ReadAttribute().
- Added comment that variables come from test plan.
- Changed to using logging.info() in place of self.print_step()

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
added spec link to vars

j-ororke committed:
updating to using WriteClient.cpp, WriteClient.h, Test_TC_ACL_2_5.yaml, and TestWriteChunking.cpp files from master to make sure that branches are in sync

j-ororke committed:
Removing Test_TC_ACL_2_7 yaml script

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,257 @@
+            result[0].Status, Status.Success, "Write should have succeeded")
+
+        self.step(7)
+        # TH1 reads Extension attribute
```
Can you add a follow up here to fix this test and the test plan as well? The test plan is not nearly specific enough here. This read is fabric filtered, which means the data for the other fabrics is filtered out, but this should also test a non-fabric filtered read that the data from the other fabric is not readable across fabrics since it's fabric sensitive.

j-ororke replied:
Created following follow-up task to enhance tests to cover testing fabric filtered and fabric sensitive scenarios:

https://github.com/project-chip/matter-test-scripts/issues/549

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,257 @@
+        events_response = await self.th1.ReadEvent(
+            self.dut_node_id,
+            events=[(0, acec_event)],
+            fabricFiltered=True
```
test plan is not specific enough here, but this should be fabricFiltered = False to ensure that the fabric sensitive information is not leaked across fabrics even when it's not filtered out. Suggest doing as a follow up.

j-ororke replied:
Understood, created the following follow-up task here to include extending testing of both the fabric filtered and fabric sensitive scenarios:
https://github.com/project-chip/matter-test-scripts/issues/549

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,257 @@
+        for event_data in events:
+            logging.info("Examining event: %s", str(event_data))
+
+            if hasattr(event_data, 'Data') and hasattr(event_data.Data, 'changeType'):
```
This step also needs to ensure the event from F2 is not included

cecille replied:
same below

j-ororke replied:
Thank you, I have added the following checks in to make sure of the following:
- For test step 9 we don't notice an entry for F2
- For test step 10 we don't notice an entry for F1

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,257 @@
+        extension_attr = Clusters.AccessControl.Attributes.Extension
+        result = await self.read_single_attribute_check_success(dev_ctrl=self.th1, endpoint=0, cluster=ac_cluster, attribute=extension_attr)
+        logging.info("TH1 read result: %s", str(result))
+        self.print_step("TH1 read result: %s", str(result))
```
```suggestion
        logging.info("TH1 read result: %s", str(result))
```

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,257 @@
+                             "Should have exactly one extension")
+
+        endpoint_data = result[0]
+        self.print_step("TH1 endpoint data: %s", str(endpoint_data))
```
```suggestion
        logging.info("TH1 endpoint data: %s", str(endpoint_data))
```

j-ororke committed:
Apply suggestions from code review from Cecille

- Changed to using logging.info() instead of self.print_step()

Co-authored-by: C Freeman <cecille@google.com>

j-ororke committed:
Update TC_ACL_2_7 python3 test module:
- Added check to make sure that for test step 9 we dont see TH2 entry and for test step 10 we dont see TH1 entry

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_7_test

Alami-Amine commented:
- Backward Compatibility: Passed Tests when run against `chip-all-clusters-app` from Matter 1.0

SamCullen-Element reviewed (dismissed):
Line 156 is a duplicate of 155, recommend removal.

andy31415 commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -0,0 +1,277 @@
+        extension_attr = Clusters.AccessControl.Attributes.Extension
+        result = await self.read_single_attribute_check_success(dev_ctrl=self.th1, endpoint=0, cluster=ac_cluster, attribute=extension_attr)
+        logging.info("TH1 read result: %s", str(result))
+        logging.info("TH1 read result: %s", str(result))
```
```suggestion
```

As per @SamCullen-Element

andy31415 committed:
Update src/python_testing/TC_ACL_2_7.py

andy31415 committed:
Merge branch 'master' into update_TC_ACL_2_7_test
